### PR TITLE
Move Field to Typescript

### PR DESCRIPTION
**Pull request by JorikSchellekens**
Moves 
- Field
- Tooltip
to Typescript

Adds @types/react-dom and @types/lodash as dev dependencies.

'Fixes' some things in QueryMatcher to allow for this.

### CODE REVIEW

File: src/autocomplete/QueryMatcher.ts [hunk 1]
```diff
@@ -75,7 +75,11 @@ export default class QueryMatcher<T> {
         this._items = new Map();
 
         for (const object of objects) {
-            const keyValues = _at(object, this._keys);
+            // Need to use unsafe coerce here because the objects can have any
+            // type for their values. We assume that those values who's keys have
+            // been specified will be string. Also, we cannot infer all the
+            // types of the keys of the objects at compile.
+            const keyValues: (string)[] = _at<T>(object as any, this._keys) as any;
```
Comment (turt2live) [inline src/autocomplete/QueryMatcher.ts:82]: `_at` is actually looking for a `string` type, not `T`. Also the array type isn't needed when specifying the type to the `_at` function:

```suggestion
            const keyValues = _at<string>(<any>object, this._keys);
```

File: src/components/views/elements/Field.tsx [hunk 2]
```diff
@@ -29,51 +29,91 @@ function getId() {
...
-        // If specified, an additional class name to apply to the field container
-        className: PropTypes.string,
-        // All other props pass through to the <input>.
-    };
+interface IProps extends React.InputHTMLAttributes<HTMLSelectElement | HTMLInputElement> {
+    // The field's ID, which binds the input and label together. Immutable.
+    id?: string,
+    // The element to create. Defaults to "input".
+    // To define options for a select, use <Field><option ... /></Field>
+    element?: InputType,
```
Comment (turt2live) [inline src/components/views/elements/Field.tsx:37]: This shouldn't be an enum and instead should be `"input" | " select" | "textarea"` (the enum doesn't add enough value here)

File: src/components/views/elements/Field.tsx [hunk 3]
```diff
@@ -29,51 +29,91 @@ function getId() {
...
+     */
+    validateOnChange = debounce(() => {
+        this.validate({
+            focused: true,
+        });
+    }, VALIDATION_THROTTLE_MS);
 
+    focus() {
+        this.input.focus();
+    }
```
Comment (turt2live) [inline src/components/views/elements/Field.tsx:110]: These should be below the constructor

File: src/components/views/elements/Field.tsx [hunk 4]
```diff
@@ -29,51 +29,91 @@ function getId() {
...
+     * We may find that we actually want different behaviour for registration
+     * fields, in which case we can add some options to control it.
+     */
+    validateOnChange = debounce(() => {
+        this.validate({
+            focused: true,
+        });
+    }, VALIDATION_THROTTLE_MS);
 
+    focus() {
```
Comment (turt2live) [inline src/components/views/elements/Field.tsx:108]: Please define `public`, `private`, or `protected` for all members.

File: src/components/views/elements/Tooltip.tsx [hunk 5]
```diff
@@ -18,67 +18,68 @@ limitations under the License.
...
+        visible?: boolean,
         // the react element to put into the tooltip
-        label: PropTypes.node,
-    },
+        label: React.ReactNode,
+}
+
+export default class Tooltip extends React.Component<IProps> {
+    private tooltipContainer: HTMLElement;
+    private tooltip: void | Element | Component<Element, any, any>;
```
Comment (turt2live) [inline src/components/views/elements/Tooltip.tsx:45]: I really hate how ReactDom's types are defined here :(

but there's nothing we can really do about it except complain.

File: src/components/views/elements/Tooltip.tsx [hunk 6]
```diff
@@ -18,67 +18,68 @@ limitations under the License.
...
+    private parent: Element;
+
 
-    getDefaultProps() {
-        return {
-            visible: true,
-        };
-    },
+    static defaultProps = {
```
Comment (turt2live) [inline src/components/views/elements/Tooltip.tsx:49]: `public`, `private`, or `protected` on all members please.

File: src/components/views/elements/Field.tsx (hunk 3 above)
Comment (JorikSchellekens) [reply src/components/views/elements/Field.tsx:110]: Having it below the constructor trips our tslint

Comment (JorikSchellekens) [reply src/components/views/elements/Field.tsx:110]: ah, focus, yes

File: src/components/views/elements/Tooltip.tsx (hunk 5 above)
Comment (JorikSchellekens) [reply src/components/views/elements/Tooltip.tsx:45]: 😢

File: src/autocomplete/QueryMatcher.ts (hunk 1 above)
Comment (JorikSchellekens) [reply src/autocomplete/QueryMatcher.ts]: Nice, much cleaner

**Comment by JorikSchellekens** [comment]
Cleaned up those bits and pieces. It'll need https://github.com/vector-im/riot-web/pull/13809 to pass the e2e tests.

**Pull request merged**